web/app/scripts/views/map/layers-controller.js

```diff
@@ -340,39 +358,38 @@
             var data = JSON.parse(record.data);
-            var startingUnderscoreRegex = /^_/;
 
             // add header with event date constant field
             /* jshint camelcase: false */
-            var str = '<div class="record-popup">';
-            str += '<div><h3>Occurred on: ' + record.occurred_from + '</h3>';
+            var str = '<div id="record-popup" class="record-popup">';
+            str += '<div><h3>' + ctl.recordTypeLabel + '</h3>';
+            // The ng-click here refers to a function which sits on the map-controller's scope
+            str += '<p>Date: <a ng-click="showDetailsModal(\'' + record.uuid + '\')">' +
+                new Date(record.occurred_from).toLocaleString() + '</a></p>';
             /* jshint camelcase: true */
 
-            // build HTML for popup from the record object
-            function strFromObj(obj) {
-                angular.forEach(obj, function(value, key) {
-                    // Skip _localId hashes, any other presumably private values
-                    // starting with an underscore, and their children.
-                    // Checking type because some keys are numeric.
-                    if (typeof key === 'string' && !key.match(startingUnderscoreRegex)) {
-                        if (typeof value === 'object') {
-                            str += '<h4>' + key + '</h4><div style="margin:15px;">';
-                            // recursively add child things, indented
-                            strFromObj(value);
-                            str += '</div>';
-                        } else {
-                            // have a simple value; display it
-                            str += '<p>' + key + ': ' + value + '</p>';
-                        }
-                    }
-                });
-            }
-
-            strFromObj(data);
-
+            // This chunk of underscore should produce an object which maps the namespace of
+            //  a set of props to the props under that namespace which are filters
+            var propsToShow = _(ctl.recordSchemaFilterables).keys()  // Only interested in which paths are relevant
+              .map(function(filterPath) { return filterPath.split('#'); })  // Split out path components
+              .groupBy(function(namespaced) { return namespaced[0]; })  // Group by the first part of the path
+              .mapValues(function(filterPathArray) {
+                  return _.map(filterPathArray, function(filterPath) { return filterPath[1]; });
+              }).value();
+
+            // Now, we need to parse the record data and crawl through it to generate a quick view
+            //  of all the filterable data on this record
+            _(propsToShow).keys().each(function(key) {
+                str += '<h4>' + key + '</h4>';
+                if (data.hasOwnProperty(key)) {
+                    _.each(propsToShow[key], function(val) {
+                        var toPrint = data[key][val] || 'UNDEFINED';
```

Comment:
I'm going to merge his PR now. Would you be able to rebase, and wire up the buttons? It seems the logic for building up the HTML will also be able to be removed.